### PR TITLE
NEW: Integrate provenance lib into the framework

### PR DESCRIPTION
**Pull request by colinvwood**

### CODE REVIEW

File: qiime2/sdk/plugin_manager.py [hunk 1]
```diff
@@ -45,10 +45,10 @@ def iter_entry_points(cls):
         for entry_point in pkg_resources.iter_entry_points(
                 group=cls.entry_point_group):
             if 'QIIMETEST' in os.environ:
-                if entry_point.name == 'dummy-plugin':
+                if entry_point.name in ('dummy-plugin', 'other-plugin'):
                     yield entry_point
             else:
-                if entry_point.name != 'dummy-plugin':
+                if entry_point.name not in ('dummy-plugin', 'other-plugin'):
```
Comment (ebolyen) [inline qiime2/sdk/plugin_manager.py:51]: :+1:

**Comment by Oddant1** [comment]
Still looking through this, but good work commenting the hell out of everything. Very helpful

File: qiime2/core/archive/provenance_lib/assets/q2_citation.bib [hunk 2]
```diff
@@ -0,0 +1,12 @@
+@Article{Bolyen2019,
```
Comment (ebolyen) [inline qiime2/core/archive/provenance_lib/assets/q2_citation.bib]: We don't need this twice, it should be the same as this one:
https://github.com/qiime2/qiime2/blob/dev/qiime2/citations.bib

File: qiime2/sdk/usage.py [hunk 3]
```diff
@@ -42,9 +42,9 @@
 
 import qiime2
 from qiime2 import sdk
-from qiime2.core.type import (is_semantic_type,
-                              is_visualization_type,
-                              is_collection_type)
+from qiime2.core.type import (
+    is_semantic_type, is_visualization_type, is_collection_type
+)
```
Comment (Oddant1) [inline qiime2/sdk/usage.py:47]: BLOATING the diff

File: qiime2/core/testing/plugin.py [hunk 4]
```diff
@@ -1088,3 +1088,32 @@ def factory():
...
+    description='',
+    short_description='',
+    version='0.0.0-dev',
+    website='',
+    package='qiime2.core.archive.provenance_lib.tests',
+    user_support_text='',
+    citations=[]
+)
+other_plugin.methods.register_function(
+    function=concatenate_ints,
```
Comment (ebolyen) [inline qiime2/core/testing/plugin.py:1104]: Thief! Stolen function! Plagiarized registration!

:+1:

File: qiime2/core/type/signature.py [hunk 5]
```diff
@@ -763,6 +764,8 @@ def _make_hashable(self, collection):
                 collection.save(fp)
                 collection = md5sum(fp)
                 return collection
+        elif isinstance(collection, MetadataInfo):
+            return collection.md5sum_hash
```
Comment (ebolyen) [inline qiime2/core/type/signature.py:768]: Thank You PyYAML, Very Cool!

File: qiime2/core/util.py [hunk 6]
```diff
@@ -122,6 +124,18 @@ def md5sum(filepath):
     return md5.hexdigest()
 
 
+def md5sum_zip(zf: zipfile.ZipFile, filepath: str) -> str:
```
Comment (ebolyen) [inline qiime2/core/util.py:127]: We can refactor these to use the same code, but that's a rainy day problem.

File: qiime2/sdk/util.py [hunk 7]
```diff
@@ -25,6 +29,20 @@
 ]
 
 
+def camel_to_snake(name: str) -> str:
```
Comment (ebolyen) [inline qiime2/sdk/util.py:32]: Is this used by q2cli?

**Review by ebolyen — APPROVED** [review]
What a doozy of a code-transfer. Great work! There's a handful of spots we could continue to DRY up, but really nothing of concern here.

File: qiime2/core/type/signature.py (hunk 5 above)
Comment (Oddant1) [reply qiime2/core/type/signature.py:768]: Colin and I had some fun with this thing lol

**Review by Oddant1 — APPROVED** [review]

File: qiime2/sdk/util.py (hunk 7 above)
Comment (colinvwood) [reply qiime2/sdk/util.py:32]: yes one test over there needs it

File: qiime2/core/archive/provenance_lib/assets/q2_citation.bib (hunk 2 above)
Comment (colinvwood) [reply qiime2/core/archive/provenance_lib/assets/q2_citation.bib]: fixed

**Review by lizgehret — APPROVED** [review]

**Pull request merged**